From: Kristian Nielsen <knielsen@knielsen-hq.org>
Date: Thu, 1 Apr 2021 21:28:00 +0100
Subject: Remove unnecessary zip_source_close

This patch fixes CVE-2021-21772, a use-after-free bug. It is a
backport of the upstream fix in v2.1.1.

Forwarded: not-needed
---
 Include/Common/OPC/NMR_OpcPackageReader.h  |  1 -
 Source/Common/OPC/NMR_OpcPackageReader.cpp | 16 ++++++----------
 2 files changed, 6 insertions(+), 11 deletions(-)

--- a/Include/Common/OPC/NMR_OpcPackageReader.h
+++ b/Include/Common/OPC/NMR_OpcPackageReader.h
@@ -54,7 +54,6 @@ namespace NMR {
 		std::vector<nfByte> m_Buffer;
 		zip_error_t m_ZIPError;
 		zip_t * m_ZIParchive;
-		zip_source_t * m_ZIPsource;
 		std::map <std::string, nfUint64> m_ZIPEntries;
 		std::map <std::string, POpcPackagePart> m_Parts;
 
diff --git a/Source/Common/OPC/NMR_OpcPackageReader.cpp b/Source/Common/OPC/NMR_OpcPackageReader.cpp
index 16dd2e8c..4f3a604d 100644
--- a/Source/Common/OPC/NMR_OpcPackageReader.cpp
+++ b/Source/Common/OPC/NMR_OpcPackageReader.cpp
@@ -111,7 +111,7 @@ namespace NMR {
 		m_ZIPError.sys_err = 0;
 		m_ZIPError.zip_err = 0;
 		m_ZIParchive = nullptr;
-		m_ZIPsource = nullptr;
+		zip_source_t* pZIPsource = nullptr;
 
 		try {
 			// determine stream size
@@ -131,20 +131,20 @@ namespace NMR {
 #endif
 			if (bUseCallback) {
 				// read ZIP from callback: faster and requires less memory
-				m_ZIPsource = zip_source_function_create(custom_zip_source_callback, pImportStream.get(), &m_ZIPError);
+				pZIPsource = zip_source_function_create(custom_zip_source_callback, pImportStream.get(), &m_ZIPError);
 			}
 			else {
 				// read ZIP into memory
 				m_Buffer.resize((size_t)nStreamSize);
 				pImportStream->readBuffer(&m_Buffer[0], nStreamSize, true);
-				m_ZIPsource = zip_source_buffer_create(&m_Buffer[0], (size_t)nStreamSize, 0, &m_ZIPError);
+				pZIPsource = zip_source_buffer_create(&m_Buffer[0], (size_t)nStreamSize, 0, &m_ZIPError);
 			}
-			if (m_ZIPsource == nullptr)
+			if (pZIPsource == nullptr)
 				throw CNMRException(NMR_ERROR_COULDNOTREADZIPFILE);
 
-			m_ZIParchive = zip_open_from_source(m_ZIPsource, ZIP_RDONLY | ZIP_CHECKCONS, &m_ZIPError);
+			m_ZIParchive = zip_open_from_source(pZIPsource, ZIP_RDONLY | ZIP_CHECKCONS, &m_ZIPError);
 			if (m_ZIParchive == nullptr) {
-				m_ZIParchive = zip_open_from_source(m_ZIPsource, ZIP_RDONLY, &m_ZIPError);
+				m_ZIParchive = zip_open_from_source(pZIPsource, ZIP_RDONLY, &m_ZIPError);
 				if (m_ZIParchive == nullptr)
 					throw CNMRException(NMR_ERROR_COULDNOTREADZIPFILE);
 				else
@@ -208,13 +208,9 @@ namespace NMR {
 		if (m_ZIParchive != nullptr)
 			zip_close(m_ZIParchive);
 
-		if (m_ZIPsource != nullptr)
-			zip_source_close(m_ZIPsource);
-
 		zip_error_fini(&m_ZIPError);
 		m_Buffer.resize(0);
 
-		m_ZIPsource = nullptr;
 		m_ZIParchive = nullptr;
 	}
 
